### PR TITLE
Add some type annotations in `synapse.storage`

### PR DESCRIPTION
**Pull request by richvdh**
I cracked, and added some type definitions in `synapse.storage`.

A few pointers here to help the reviewer:

 * `synapse.storage.types.{Cursor,Connection}` are defined as interfaces which are implemented by the sqlite3 and psycopg2 implementations
 * `LoggingTransaction` implicitly also implements `Cursor`.
 * `BaseDatabaseEngine` is an abstract base class which is implemented by `Sqlite3Engine` and `PostgresEngine`.
 * `create_engine` is rewritten with less magic, to keep mypy happy
 * ditto `LoggingTransaction`.
 * a bunch of other cleanups and annotations to appease mypy.

### CODE REVIEW

File: synapse/storage/database.py [hunk 1]
```diff
@@ -107,20 +99,25 @@ def make_conn(db_config: DatabaseConnectionConfig, engine):
     return db_conn
 
 
-class LoggingTransaction(object):
+# the type of entry which goes on our after_callbacks and exception_callbacks lists
+# use of a string here because python 3.5.2 doesn't support Callable([...]).
```
Comment (babolivier) [inline synapse/storage/database.py]: Hmm, I think I sort of understand this comment, but not really sure and I had to read it a few times to get there, could you maybe rephrase it?

File: synapse/storage/database.py [hunk 2]
```diff
@@ -135,46 +132,61 @@ class LoggingTransaction(object):
...
+        self.after_callbacks = after_callbacks
+        self.exception_callbacks = exception_callbacks
 
-    def call_after(self, callback, *args, **kwargs):
+    def call_after(self, callback: "Callable[..., None]", *args, **kwargs):
         """Call the given callback on the main twisted thread after the
         transaction has finished. Used to invalidate the caches on the
         correct thread.
         """
+        assert self.after_callbacks is not None
```
Comment (babolivier) [inline synapse/storage/database.py]: Not sure I understand why we're raising an exception here if `self.after_callbacks` is None and not assigning a new list in that case.

File: synapse/storage/engines/postgres.py [hunk 3]
```diff
@@ -15,16 +15,14 @@
...
-from ._base import IncorrectDatabaseSetup
+from ._base import BaseDatabaseEngine, IncorrectDatabaseSetup
 
 logger = logging.getLogger(__name__)
 
 
-class PostgresEngine(object):
-    single_threaded = False
-
+class PostgresEngine(BaseDatabaseEngine):
```
Comment (babolivier) [inline synapse/storage/engines/postgres.py]: Isn't `BaseDatabaseEngine` missing a connection type? (seeing as `Sqlite3Engine` has `BaseDatabaseEngine[sqlite3.Connection]`)

File: synapse/storage/types.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+from typing_extensions import Protocol
+
+
+"""
+Some very basic protocol definitions for the DB-API2 classes specified in PEP-249
+"""
+
+
+class Row(Protocol):
+    # todo: make this stronger
```
Comment (babolivier) [inline synapse/storage/types.py]: "stronger" is pretty vague, can you elaborate here please?

File: synapse/storage/types.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
+# -*- coding: utf-8 -*-
```
Comment (babolivier) [inline synapse/storage/types.py]: Aren't the classes described in this file supposed to be abstract classes?

File: synapse/storage/database.py (hunk 2 above)
Comment (richvdh) [reply synapse/storage/database.py]: well, it's up to the thing that constructs the `LoggingTransaction` to make sure that anything that gets added to `after_callbacks` actually gets called. So if we just create an `after_callbacks` ourselves, then the callback will just get silently dropped.

In short, if anything calls `call_after` when `after_callbacks is None`, that means there's a programming error.

No doubt we could factor out the code to run `after_callbacks` to avoid the problem, but this PR was about annotating the current impl rather than trying to change too much.

have added some comments which might help explain.

File: synapse/storage/engines/postgres.py (hunk 3 above)
Comment (richvdh) [reply synapse/storage/engines/postgres.py]: well, I don't think it's mandatory, and the problem is that there isn't an equivalent connection type for Postgres.

File: synapse/storage/types.py (hunk 4 above)
Comment (richvdh) [reply synapse/storage/types.py]: actually I'll just get rid of `Row` and replace it with `Tuple`.

File: synapse/storage/types.py (hunk 5 above)
Comment (richvdh) [reply synapse/storage/types.py]: well, they are `Protocol`s. There's a whole bunch of stuff about that at https://mypy.readthedocs.io/en/stable/protocols.html, but basically it's a way of defining an interface that is implemented by other classes, rather than anything that is meant to be used as a base class for anything else.

Comment (babolivier) [reply synapse/storage/types.py]: Oooh TIL.

File: synapse/storage/database.py (hunk 2 above)
Comment (babolivier) [reply synapse/storage/database.py]: Right, fair enough, fwiw I think the comment is enough.

File: synapse/storage/engines/postgres.py (hunk 3 above)
Comment (babolivier) [reply synapse/storage/engines/postgres.py]: Ah, fair enough then
